flatpak_builder_lint/checks/appid.py

```diff
@@ -1,17 +1,42 @@
 import os
+import re
+import tempfile
 from typing import Optional
 
-from .. import builddir
+from .. import builddir, domainutils, ostree
 from . import Check
 
 
 class AppIDCheck(Check):
-    def _validate(self, appid: Optional[str]) -> None:
+    def _validate(self, appid: Optional[str], is_extension: bool) -> None:
         if not appid:
             self.errors.add("appid-not-defined")
             return
 
+        if len(appid) > 255:
+            self.errors.add("appid-length-more-than-255-chars")
+            return
+
         split = appid.split(".")
+
+        if len(split) < 3:
+            self.errors.add("appid-less-than-3-components")
+            return
+
+        if not all(re.match("^[A-Za-z_][\\w\\-]*$", sp) for sp in split):
+            self.errors.add("appid-component-wrong-syntax")
+            return
+
+        is_baseapp = appid.endswith(".BaseApp")
+
+        if not (is_extension or is_baseapp) and len(split) >= 6:
+            self.errors.add("appid-too-many-components-for-app")
```

Comment:
Only two "apps" fail this:

```
uk.org.greenend.chiark.sgtatham.putty
uk.org.greenend.chiark.sgtatham.puzzles
```

I think we should restrict this, because if not here, they will fail at one of the domain checks. Supporting arbitrarily many components for apps is complicated.

I don't think anything in the near future will require more than 5 components.